### ISSUE TITLE
Issues using coreclr on Linux without Mono

### ISSUE THREAD

**Issue by rlordcardano** [pythonnet/pythonnet#801]
### Environment

-   Pythonnet version: PR #612 
-   Python version: 3.6.7
-   Operating System: Ubuntu 18.04

### Details

I was trying to test PR #612 for our use case (calling functionality implemented in .NET Standard 2.0 DLLs from Python 3.x). We currently manage to do this on Linux by using Mono and Pythonnet, but we are trying to avoid the dependency on Mono going forward.

[Here](https://gist.github.com/rlordcardano/c1c961effec6be2ddeab8798c0e85c29) I have included the dockerfile used to setup the required environment (Ubuntu 18.04 docker image, with .NET Core SDK, Python 3.6.7 and a clone of PR #612) so that one can reproduce what I'm seeing.

Subsequently, after starting the docker image, run:

```
cd pythonnet
python3 setup.py bdist_wheel --xplat
```

Subsequently, install the wheel that has been generated in the dist folder.
The error I get after this is:

```
>>>import clr
Failed to convert CLR files path to absolute path
Traceback (most recent call shown last):
  File "<stdin>", line 1, in <module>
ImportError: Unable to find clr path
```

**Comment by denfromufa** [comment]
does npython.exe work for you? you can build it from the full solution in pythonnet repository.

**Comment by rlordcardano** [comment]
That could work for one set of the use cases (I see now in #96 that the only embedding that is supported on *nix is via nPython, so I take it the other route doesn't work yet), the other set is using our C# DLLs within Jupyter notebooks - I don't think nPython.exe would be used there, would it?

Trying to test nPython.exe in my container `dotnet nPython.dll` gives:

```
Unhandled Exception: System.DllNotFoundException: Unable to load shared library 'python2.7m' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: libpython2.7m: cannot open shared object file: No such file or directory
   at Python.Runtime.Runtime.Py_IsInitialized()
   at Python.Runtime.Runtime.Initialize(Boolean initSigs) in /pythonnet/src/runtime/runtime.cs:line 274
   at Python.Runtime.PythonEngine.Initialize(IEnumerable`1 args, Boolean setSysArgv, Boolean initSigs) in /pythonnet/src/runtime/pythonengine.cs:line 169
   at Python.Runtime.PythonConsole.Main(String[] args) in /pythonnet/src/console/pythonconsole.cs:line 36
Aborted
```

I'm sure I'm missing some basic setting to point it to my Python 3.x installation, any help welcome though...

**Comment by denfromufa** [comment]
PR #612 was supposed to add clr.so (clr.pyd) library that makes it possible to `import clr` pythonnet from python without npython.exe.

I asked you to check npython.exe to make sure that the problem comes from PR #612, which requires debugging now.

**Comment by rlordcardano** [comment]
@denfromufa to be sure, nPython.exe does not work:

`A fatal error was encountered. The library 'libhostpolicy.so' required to execute the application was not found in '/pythonnet/src/console/bin/'.
Failed to run as a self-contained app. If this should be a framework-dependent app, add the /pythonnet/src/console/bin/nPython.runtimeconfig.json file specifying the appropriate framework.`

but I wouldn't expect it to as it targets .NET 4.0 (presumably it would work after installing Mono and the required runtime). The netcoreapp2.0 target framework produces nPython.dll, which does work.

**Comment by denfromufa** [comment]
@rlordcardano yes, I should have been more specific - most likely `nPython.exe` is a .NET Framework or Mono build (depending on platform), while `nPython.dll` is a .NET Core build.

**Comment by rlordcardano** [comment]
My two cents:

* **[/src/coreclr/pynetinit.c](https://github.com/djoyce82/pythonnet/blob/master/src/coreclr/pynetinit.c#L171)** - line 171 states:

`    if (!GetClrFilesAbsolutePath(pn_args->entry_path, "/usr/share/dotnet/shared/Microsoft.NETCore.App/2.0.0", &pn_args->clr_path))`

This is not always the location of the .NET Core runtime, e.g. on my system it's 2.1.5. There must be a way to find out programmatically what version of the runtime is installed. E.g. `dotnet --info` gives the required information. This sounds like the easy part.

* **[/src/runtime.pythonengine.cs](https://github.com/djoyce82/pythonnet/blob/master/src/runtime/pythonengine.cs#L168)**: After setting that to the right path, it eventually crashes in line 168 (Runtime.Initialize). Error message "free(): invalid pointer".

Haven't gone any further yet, but maybe this will help for others that are looking at it.

**Comment by Cronan** [comment]
On CentOS 7 the locations are not very consistent:

* /usr/share/dotnet/shared/Microsoft.NETCore.App/2.0.0
* /opt/rh/rh-dotnet20/root/lib64/dotnet/shared/Microsoft.NETCore.App/2.0.7
* /opt/rh/rh-dotnet21/root/usr/lib64/dotnet/shared/Microsoft.NETCore.App/2.1.2

I'm not sure probing for the location will be very successful.
One way to solve it might be with an env variable, like so:

```c
static const char* coreRootVar = "CORE_ROOT";
const char* clrFilesPath = getenv(coreRootVar);
if (clrFilesPath == NULL) {
    clrFilesPath = "/usr/share/dotnet/shared/Microsoft.NETCore.App/2.0.0";
}

if (!GetClrFilesAbsolutePath(pn_args->entry_path, clrFilesPath, &pn_args->clr_path))
{
    pn_args->error = "Unable to find clr path";
    return;
}
```

I chose **CORE_ROOT** because it's [already used](https://github.com/dotnet/coreclr/blob/master/Documentation/workflow/UsingCoreRun.md) by **corerun.exe**.

This could be set in python or in the shell.

**Comment by filmor** [comment]
I'm looking into a pure Python solution using ctypes right now, there I'm parsing the output of the `dotnet` command to find the clr files.

**Comment by filmor** [comment]
Closed in favour of #984.
